contrib/epee/src/net_ssl.cpp

```diff
@@ -522,30 +522,101 @@ bool ssl_options_t::handshake(
       return true;
     });
   }
+}
+
+bool ssl_options_t::handshake(
+  boost::asio::ssl::stream<boost::asio::ip::tcp::socket> &socket,
+  boost::asio::ssl::stream_base::handshake_type type,
+  boost::asio::const_buffer buffer,
+  const std::string& host,
+  std::chrono::milliseconds timeout) const
+{
+  configure(socket, type, host);
+
+  auto start_handshake = [&]{
+    using ec_t = boost::system::error_code;
+    using timer_t = boost::asio::steady_timer;
+    using strand_t = boost::asio::io_service::strand;
+    using lock_t = std::mutex;
+    using lock_guard_t = std::lock_guard<lock_t>;
+    using condition_t = std::condition_variable_any;
+    using socket_t = boost::asio::ip::tcp::socket;
+
+    auto &io_context = GET_IO_SERVICE(socket);
+    if (io_context.stopped())
+      io_context.reset();
+    strand_t strand(io_context);
+    timer_t deadline(io_context, timeout);
+
+    struct state_t {
+      lock_t lock;
```

Comment:
Did you check the `strand_deadlock` test case? I'm quite sure some of these design decisions were intentional to solve this issue, together with setting `-DBOOST_ASIO_ENABLE_SEQUENTIAL_STRAND_ALLOCATION`

But I also don't want to give false info so maybe ignore if this doesn't make sense.

Reply:
Yes I'm sure the author put the locks in there for a reason, due to the current design of `handshake`, but since this is a complete rewrite,  it doesn't make sense to have both code running asynchronously from a single executor AND have status booleans AND have it be manually, externally locked. Even if it functions correctly, this will likely gum up the `io_context` of `socket` waiting for locks and it makes it a ton harder to track what's going on.